File: C#/Loon2MonoGame/LoonMonoGame-Lib/loon/utils/reply/Closeable.cs
using java.lang;
using java.util;

namespace loon.utils.reply
{

    public abstract class Closeable : LRelease
    {
        internal class Set : Closeable
        {
            protected ObjectSet<LRelease> _set;

            protected bool _closed = false;

            public bool IsClosed()
            {
                return _closed;
            }

            public override void Close()
            {
                if (_set != null)
                {
                    for (Iterator<LRelease> it=_set.Iterator(); it.HasNext();)
                    {
                        LRelease c = it.Next();
                        try
                        {
                            c.Close();
                        }
                        catch (System.Exception)
                        {
                        }
                    }
                    _set.Clear();
                }
                _closed = true;
            }

            public T Add<T>(T c) where T : LRelease
            {
               if (_set == null)
                {
                    _set = new ObjectSet<LRelease>();
                }
                _set.Add(c);
                return c;
            }

            public void Remove(LRelease c)
            {
                if (_set != null)
                {
                    _set.Remove(c);
                }
            }
        }

        public abstract void Close();
    }

}

File: Java/Loon-Neo/src/loon/utils/reply/Closeable.java
/**
 * Copyright 2008 - 2015 The Loon Game Engine Authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.5
 */
package loon.utils.reply;

import loon.LRelease;
import loon.utils.ObjectSet;

public interface Closeable extends LRelease {

	class Set implements Closeable {

		protected ObjectSet<LRelease> _set;

		protected boolean _closed = false;

		public boolean isClosed() {
			return _closed;
		}

		@Override
		public void close() {
			if (_set != null) {
				for (LRelease c : _set) {
					try {
						c.close();
					} catch (Throwable e) {
					}
				}
				_set.clear();
			}
			_closed = true;
		}

		public <T extends LRelease> T add(T c) {
			if (_set == null) {
				_set = new ObjectSet<LRelease>();
			}
			_set.add(c);
			return c;
		}

		public void remove(LRelease c) {
			if (_set != null) {
				_set.remove(c);
			}
		}
	}
	
	@Override
    public abstract void close();

}
